### ISSUE TITLE
refactor EP inference in GP

### ISSUE THREAD

**Issue by yorkerlin** [shogun-toolbox/shogun#3407]
@karlnapf 
The existing EP inference is not scalable.
This issue is about refactoring existing EP inference in Shogun and extend it to [FITC EP](http://www.gaussianprocess.org/gpml/code/matlab/inf/infFITC_EP.m) inference.

**Comment by yorkerlin** [comment]
I will do it soon. Hopefully, the method will be added in this week.

**Comment by yorkerlin** [comment]
It is also important to check whether there is a bug in the existing EP inference when it comes to non-zero mean function. The existing unit tests only cover cases of zero mean.

**Comment by karlnapf** [comment]
Great, is there a benchmark? Not scalable in what sense?

**Comment by karlnapf** [comment]
Will be away for some time but answer when back in Sept

**Comment by yorkerlin** [comment]
The existing EP implementation is for `full` GP.
EP under FITC approximation is for sparse GP.

**Comment by karlnapf** [comment]
Ah I see...so we dont have these fancy stochastic EP updates for large scale sparse classification

**Comment by yorkerlin** [comment]
No, it is about `batch` EP inference under FITC approximation.

**Comment by yorkerlin** [comment]
For stochastic inference, determining convergence is the main issue. 
Users have to select right step sizes or learning rate. 


For black box usage, (hold-out) evaluation data set is used to determine convergence. It is also true in deep neural network. For black box usage, It is hard to determine convergence without using evaluation data set.

**Comment by yorkerlin** [comment]
There are two EP methods for sparse GP. I will add them this year.

**Comment by yorkerlin** [comment]
I and Emt are working on a paper related to EP.

**Comment by yorkerlin** [comment]
For reference:
* method one from paper, https://papers.nips.cc/paper/3351-the-generalized-fitc-approximation.pdf
* method two from paper, https://arxiv.org/abs/1507.04513
